### PR TITLE
Airmode 3D final fix // Tested

### PR DESCRIPTION
**Pull request by borisbstyle**
@hydra 
Finally all tests completed. 3D mode works as intended now! Sorry that it took a while to test it.

### CODE REVIEW

**Comment by borisbstyle** [comment]
@hydra 
This can be merged! Had to change rcCommand with rcData. rcCommand has too much manipulation....min_check etc..

File: src/main/flight/mixer.c [hunk 1]
```diff
@@ -833,7 +829,7 @@ void mixTable(void)
             if (isFailsafeActive) {
                 mixConstrainMotorForFailsafeCondition(i);
             } else if (feature(FEATURE_3D)) {
-                if (throttle >= flight3DConfig->deadband3d_high)  {
+                if (throttlePrevious >= (rxConfig->midrc + flight3DConfig->deadband3d_throttle)) {
```
Comment (ledvinap) [inline src/main/flight/mixer.c]: You are using opposite test than on https://github.com/cleanflight/cleanflight/pull/1792/files#diff-95e70d63cf1e155a276b2ad2a8e7a947R795.
It is not problem now (throttlePrevious is always outside deadband3d_throttle), but can cause problems if code is changed

File: src/main/flight/mixer.c [hunk 2]
```diff
@@ -779,29 +780,24 @@ void mixTable(void)
         int16_t rollPitchYawMixRange = rollPitchYawMixMax - rollPitchYawMixMin;
         int16_t throttleRange, throttle;
         int16_t throttleMin, throttleMax;
+        static int16_t throttlePrevious = 0;   // Store the last throttle direction for deadband transitions in 3D
```
Comment (ledvinap) [inline src/main/flight/mixer.c]: Throttle will start in negative range with this initialization `throttlePrevious <= (rxConfig->midrc - flight3DConfig->deadband3d_throttle)`

**Comment by ledvinap** [comment]
The rcData/rcCommand should be fixed sometimes ... similar calculation is duplicated in multiple places

File: src/main/flight/mixer.c (hunk 2 above)
Comment (borisbstyle) [inline src/main/flight/mixer.c]: fair enough.

I reversed the logic there and changed initialisation to mid_rc.

The arming is allowed in both directions, but starts from positive when stick was centered. Anyway it depends how the stick was moving prior arming.

**Comment by borisbstyle** [comment]
using rcCommand in 3D is totally a bad idea for now.

But do agree 200% that rc code should somehow be cleaned up. It's all over the place and manipulated on many places!

**Comment by borisbstyle** [comment]
wait I will fix it to only be able to arm to positive as the stick moves prior arming

**Comment by borisbstyle** [comment]
Finilised testing. I think it is done now.

Added documentation!

**Comment by bdoiron74** [comment]
Brain fart - Does it not make perfect sense to have two sets of PID gains for 3D flying? Unless someone makes 3D symmetrical props for all, we need gains for two different thrust profiles.

**Comment by borisbstyle** [comment]
there are already fully symmetrical props available.
i have them from hq.
Op 14 feb. 2016 05:17 schreef "bdoiron74" <notifications@github.com>:

**Comment by bdoiron74** [comment]
I was thinking more for the pilot who wants efficient forward thrust with the occasional trick move.

**Pull request merged**